### PR TITLE
CCMSG-966: Implement data stream write for Elasticsearch

### PR DESCRIPTION
**Pull request by amuamushu**
## Problem
Elasticsearch connector does not allow writes to data streams.

## Solution
Extended functionality to alternatively allow writes to data stream if the data stream configurations are set.

##### Does this solution apply anywhere else?
- [ ] yes
- [x] no

##### Testing done:
- [x] Unit tests
- [x] Integration tests
- [ ] System tests
- [ ] Manual tests

## Release Plan

Will be released as part of the data stream feature.

Not backporting or merging to master.
 
Not reverting or rolling back.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -38,7 +38,7 @@
     </scm>
 
     <properties>
-        <es.version>7.0.1</es.version>
+        <es.version>7.12.0</es.version>
```
Comment (levzem) [inline pom.xml:41]: thought it was `7.9.x`?

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java [hunk 2]
```diff
@@ -185,6 +187,9 @@ public boolean createIndex(String index) {
     if (indexExists(index)) {
       return false;
     }
+    if (config.isDataStream()) {
+      return createDataStream(index);
+    }
 
     CreateIndexRequest request = new CreateIndexRequest(index);
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java:194]: how about we move this to another `private boolean createIndex(String index)` method and then have this method called `public boolean createIndexOrDataStream(String name)`

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java [hunk 3]
```diff
@@ -157,6 +157,48 @@ private String convertTopicToIndexName(String topic) {
...
+  /**
+   * Returns the converted index name from a given topic name in the form {type}-{dataset}-{topic}.
+   * For the <code>topic</code>, Elasticsearch accepts:
+   * <ul>
+   *   <li>all lowercase</li>
+   *   <li>no longer than 100 bytes</li>
+   * </ul>
+   * (<a href="https://github.com/elastic/ecs/blob/master/rfcs/text/0009-data_stream-fields.md#restrictions-on-values">ref</a>_.)
+   */
+  private String convertUsingIndexTemplate(String topic) {
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java:169]:
```suggestion
  private String convertTopicToDataStreamName(String topic) {
```
consistent naming

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java [hunk 4]
```diff
@@ -157,6 +157,48 @@ private String convertTopicToIndexName(String topic) {
...
+   * accepts:
+   * <ul>
+   *   <li>all lowercase</li>
+   *   <li>less than 256 bytes</li>
+   *   <li>does not start with - or _</li>
+   *   <li>is not . or ..</li>
+   * </ul>
+   * (<a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-create-index.html#indices-create-api-path-params">ref</a>_.)
+   */
+  private String createIndexNameWith(String topic) {
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java:195]:
```suggestion
  private String createIndexName(String topic) {
```

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java [hunk 5]
```diff
@@ -157,6 +157,48 @@ private String convertTopicToIndexName(String topic) {
...
+    if (config.isDataStream()) {
+      return convertUsingIndexTemplate(topic);
+    }
+    return convertTopicToIndexName(topic);
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java:199]:
```suggestion
    return config.isDataStream() ? convertUsingIndexTemplate(topic) : convertTopicToIndexName(topic);
```

Comment (levzem) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java:199]: can use a fancy ternary operator here

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java [hunk 6]
```diff
@@ -157,6 +157,48 @@ private String convertTopicToIndexName(String topic) {
     return index;
   }
 
+  /**
+   * Returns the converted index name from a given topic name in the form {type}-{dataset}-{topic}.
+   * For the <code>topic</code>, Elasticsearch accepts:
+   * <ul>
+   *   <li>all lowercase</li>
+   *   <li>no longer than 100 bytes</li>
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java:165]: the whole data stream name is less than 100 bytes or just the namespace/topic portion?

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java [hunk 7]
```diff
@@ -157,6 +157,48 @@ private String convertTopicToIndexName(String topic) {
...
+   * </ul>
+   * (<a href="https://github.com/elastic/ecs/blob/master/rfcs/text/0009-data_stream-fields.md#restrictions-on-values">ref</a>_.)
+   */
+  private String convertUsingIndexTemplate(String topic) {
+    topic = topic.toLowerCase();
+    if (topic.length() > 100) {
+      topic = topic.substring(0, 100);
+    }
+    String index = String.format(
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java:174]:
```suggestion
    String dataStream = String.format(
```

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java [hunk 8]
```diff
@@ -308,6 +310,45 @@ public void testConvertTopicToIndexName() {
     verify(client, times(1)).createIndex(eq("dotdot"));
   }
 
+  @Test
+  public void testConvertTopicToIndexTemplate() {
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java:314]: would be good to break this apart into separate tests for each test case

File: src/test/java/io/confluent/connect/elasticsearch/helper/ElasticsearchHelperClient.java [hunk 9]
```diff
@@ -49,17 +54,30 @@ public ElasticsearchHelperClient(String url, ElasticsearchSinkConnectorConfig co
...
+    if (isDataStream) {
+      DeleteDataStreamRequest request = new DeleteDataStreamRequest(index);
+      client.indices().deleteDataStream(request, RequestOptions.DEFAULT);
+      return;
+    }
     DeleteIndexRequest request = new DeleteIndexRequest(index);
     client.indices().delete(request, RequestOptions.DEFAULT);
   }
 
+  public DataStream getDataStream(String index) throws IOException {
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/helper/ElasticsearchHelperClient.java:67]:
```suggestion
  public DataStream getDataStream(String dataStream) throws IOException {
```

File: src/test/java/io/confluent/connect/elasticsearch/helper/ElasticsearchHelperClient.java [hunk 10]
```diff
@@ -49,17 +54,30 @@ public ElasticsearchHelperClient(String url, ElasticsearchSinkConnectorConfig co
     );
   }
 
-  public void deleteIndex(String index) throws IOException {
+  public void deleteIndex(String index, boolean isDataStream) throws IOException {
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/helper/ElasticsearchHelperClient.java:57]: maybe just have a `deleteDataStream()` method instead?

File: src/test/java/io/confluent/connect/elasticsearch/integration/ElasticsearchConnectorBaseIT.java [hunk 11]
```diff
@@ -146,13 +157,13 @@ protected void waitForRecords(int numRecords) throws InterruptedException {
 
   protected void writeRecords(int numRecords) {
     for (int i = 0; i < numRecords; i++) {
-      connect.kafka().produce(TOPIC, String.valueOf(i), String.format("{\"doc_num\":%d}", i));
+      connect.kafka().produce(TOPIC, String.valueOf(i), String.format("{\"doc_num\":%d,\"@timestamp\":%s}", i, "\"2021-04-28T11:11:22.000Z\""));
     }
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/integration/ElasticsearchConnectorBaseIT.java:161]: idk why i wrote it like this but lets change all of this to call `writeRecordsFromStartIndex(0, numRecords);`

File: src/test/java/io/confluent/connect/elasticsearch/integration/ElasticsearchConnectorBaseIT.java [hunk 12]
```diff
@@ -146,13 +157,13 @@ protected void waitForRecords(int numRecords) throws InterruptedException {
...
   }
 
   protected void writeRecordsFromStartIndex(int start, int numRecords) {
     for (int i  = start; i < start + numRecords; i++) {
-      connect.kafka().produce(TOPIC, String.valueOf(i), String.format("{\"doc_num\":%d}", i));
+      connect.kafka().produce(TOPIC, String.valueOf(i), String.format("{\"doc_num\":%d,\"@timestamp\":%s}", i, "\"2021-04-28T11:11:22.000Z\""));
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/integration/ElasticsearchConnectorBaseIT.java:166]:
```suggestion
      connect.kafka().produce(TOPIC, String.valueOf(i), String.format("{\"doc_num\":%d,\"@timestamp\":\"2021-04-28T11:11:22.%03dZ\"}", i, i));
```
would be nice to at least change the timestamp a bit

File: src/test/java/io/confluent/connect/elasticsearch/integration/ElasticsearchConnectorIT.java [hunk 13]
```diff
@@ -102,6 +104,18 @@ public void testHappyPath() throws Exception {
...
+  @Test
+  public void testHappyPathDataStream() throws Exception {
+    isDataStream = true;
+    props.put(DATA_STREAM_TYPE_CONFIG, "logs");
+    props.put(DATA_STREAM_DATASET_CONFIG, "dataset");
+    index = "logs-dataset-" + TOPIC;
+
+    runSimpleTest(props);
+
+    assertEquals(index, helperClient.getDataStream(index).getName());
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/integration/ElasticsearchConnectorIT.java:116]: we need to `verifySearchResults()` here too

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java (hunk 6 above)
Comment (amuamushu) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java:165]: Just the namespace/topic portion. I only mentioned the topic because it's the only parameter and the other restrictions are validated in the configs. Should I update the method header to talk about the other fields too?

File: src/test/java/io/confluent/connect/elasticsearch/integration/ElasticsearchConnectorIT.java (hunk 13 above)
Comment (amuamushu) [reply src/test/java/io/confluent/connect/elasticsearch/integration/ElasticsearchConnectorIT.java:116]: I believe `verifySearchResults()` is called in `runSimpleTest()`. Would I need to explicitly call it again?

File: src/test/java/io/confluent/connect/elasticsearch/helper/ElasticsearchHelperClient.java (hunk 10 above)
Comment (amuamushu) [reply src/test/java/io/confluent/connect/elasticsearch/helper/ElasticsearchHelperClient.java:57]: In that case, should I add a public method called `deleteIndexOrDatastream()`? Otherwise, I would have to add an if/else block wherever `deleteIndex()` is currently called, which is a bit repetitive.

**Review by amuamushu — COMMENTED** [review]
Thanks Lev for the speedy review!

File: pom.xml (hunk 1 above)
Comment (levzem) [reply pom.xml:41]: yes lets use the latest 7.9.x version

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java (hunk 2 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java:194]: IMO should be pretty self-explanatory, especially when reading the code

File: src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java (hunk 6 above)
Comment (levzem) [reply src/main/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTask.java:165]: nope its okay, I was just clarifying for myself

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java [hunk 14]
```diff
@@ -273,6 +275,81 @@ public void testVersion() {
     assertTrue(task.version().matches("^(\\d+\\.){2}?(\\*|\\d+)(-.*)?$"));
   }
 
+  @Test
+  public void testConvertTopicToDataStream_allowDashes() {
+    final String type = "logs";
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java:280]: don't need `final` anywhere

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java [hunk 15]
```diff
@@ -273,6 +275,81 @@ public void testVersion() {
     assertTrue(task.version().matches("^(\\d+\\.){2}?(\\*|\\d+)(-.*)?$"));
   }
 
+  @Test
+  public void testConvertTopicToDataStream_allowDashes() {
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java:279]:
```suggestion
  public void testConvertTopicToDataStreamAllowsDashes() {
```
dont need underscore anywhere

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java [hunk 16]
```diff
@@ -273,6 +275,81 @@ public void testVersion() {
...
+    final String type = "logs";
+    final String dataset = "a_valid_dataset";
+    props.put(DATA_STREAM_TYPE_CONFIG, type);
+    props.put(DATA_STREAM_DATASET_CONFIG, dataset);
+    setUpTask();
+
+    String startsWithDash = "-dash";
+    SinkRecord record = record(startsWithDash, true, false, 0);
+    task.put(Collections.singletonList(record));
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java:288]:
```suggestion
    task.put(Collections.singletonList(record(startsWithDash, true, false, 0)));
```

can do that everywhere

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java [hunk 17]
```diff
@@ -273,6 +275,81 @@ public void testVersion() {
...
 
+  @Test
+  public void testConvertTopicToDataStream_allowDashes() {
+    final String type = "logs";
+    final String dataset = "a_valid_dataset";
+    props.put(DATA_STREAM_TYPE_CONFIG, type);
+    props.put(DATA_STREAM_DATASET_CONFIG, dataset);
+    setUpTask();
+
+    String startsWithDash = "-dash";
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java:286]:
```suggestion
    String topic = "-dash";
```

everywhere

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java [hunk 18]
```diff
@@ -273,6 +275,81 @@ public void testVersion() {
...
+    SinkRecord record = record(startsWithDash, true, false, 0);
+    task.put(Collections.singletonList(record));
+    String indexName = convertToIndexTemplate(type, dataset, startsWithDash);
+    verify(client, times(1)).createIndex(eq(indexName));
+  }
+
+  @Test
+  public void testConvertTopicToDataStream_allowEmpty() {
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchSinkTaskTest.java:294]: topics will never be empty

File: src/test/java/io/confluent/connect/elasticsearch/integration/ElasticsearchConnectorBaseIT.java [hunk 19]
```diff
@@ -53,8 +53,11 @@
 
   protected static ElasticsearchContainer container;
 
+  protected boolean isDataStream;
   protected ElasticsearchHelperClient helperClient;
   protected Map<String, String> props;
+  protected String index;
+
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/integration/ElasticsearchConnectorBaseIT.java:60]: extra new line

File: src/main/java/io/confluent/connect/elasticsearch/DataConverter.java [hunk 20]
```diff
@@ -130,21 +131,10 @@ private String convertKey(Schema keySchema, Object key) {
           break;
         case FAIL:
         default:
-          throw new DataException(
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/DataConverter.java:133]: not necessary change for this PR, if we want to do a refactor, let's go in a separate PR and target `11.0.x`

File: src/main/java/io/confluent/connect/elasticsearch/DataConverter.java [hunk 21]
```diff
@@ -155,15 +145,21 @@ private String convertKey(Schema keySchema, Object key) {
     }
 
     // index
+    return createIndexRequest(record, index, id);
```
Comment (levzem) [inline src/main/java/io/confluent/connect/elasticsearch/DataConverter.java:148]: reason for breaking apart?

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchClientTest.java [hunk 22]
```diff
@@ -106,7 +106,7 @@ public void setup() {
   @After
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchClientTest.java:106]: lets add a test for creating a datastream and for creating a datastream that already exists

**Review by levzem — COMMENTED** [review]
almost there

File: src/main/java/io/confluent/connect/elasticsearch/DataConverter.java (hunk 21 above)
Comment (amuamushu) [reply src/main/java/io/confluent/connect/elasticsearch/DataConverter.java]: Resolved in chat. Reverted this change.

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchClientTest.java [hunk 23]
```diff
@@ -145,20 +146,53 @@ public void testCreateIndex() throws IOException {
...
+  public void testCreateNewDataStream() throws Exception {
+    String type = "logs";
+    String dataset = "dataset";
+    String dataStream = String.format("%s-%s-%s", type, dataset, INDEX);
+    props.put(DATA_STREAM_TYPE_CONFIG, type);
+    props.put(DATA_STREAM_DATASET_CONFIG, dataset);
+    config = new ElasticsearchSinkConnectorConfig(props);
+    ElasticsearchClient client = new ElasticsearchClient(config, null);
+
+    client.createIndexOrDataStream(dataStream);
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchClientTest.java:181]:
```suggestion
    assertTrue(client.createIndexOrDataStream(dataStream));
```

File: src/test/java/io/confluent/connect/elasticsearch/ElasticsearchClientTest.java [hunk 24]
```diff
@@ -145,20 +146,53 @@ public void testCreateIndex() throws IOException {
...
+  public void testCreateExistingDataStream() throws Exception {
+    String type = "logs";
+    String dataset = "dataset";
+    String dataStream = String.format("%s-%s-%s", type, dataset, INDEX);
+    props.put(DATA_STREAM_TYPE_CONFIG, type);
+    props.put(DATA_STREAM_DATASET_CONFIG, dataset);
+    config = new ElasticsearchSinkConnectorConfig(props);
+    ElasticsearchClient client = new ElasticsearchClient(config, null);
+
+    client.createIndexOrDataStream(dataStream);
```
Comment (levzem) [inline src/test/java/io/confluent/connect/elasticsearch/ElasticsearchClientTest.java:164]:
```suggestion
    assertTrue(client.createIndexOrDataStream(dataStream));
```

**Review by levzem — APPROVED** [review]
LGTM
